Model card for mobilenet v1 and v2 (#37948)
huggingface/transformers: docs/source/en/model_doc/mobilenet_v2.md

-->

<div style="float: right;">
    <div class="flex flex-wrap space-x-1">
        <img alt="PyTorch" src="https://img.shields.io/badge/PyTorch-EE4C2C?style=flat&logo=pytorch&logoColor=white">
    </div>
</div>

# MobileNet V2

> MobileNet V2 builds upon MobileNet V1, improving performance on mobile and embedded devices. Its key innovations are the **inverted residual blocks** (where shortcuts connect thin bottleneck layers) and the use of **linear bottlenecks** (removing non-linearities in narrow layers to preserve information). Like V1, it uses depthwise separable convolutions for efficiency and offers tunable hyperparameters.

Reviewer: ```suggestion
[MobileNet V2](https://huggingface.co/papers/1801.04381) improves performance on mobile devices with a more efficient architecture. It uses inverted residual blocks and linear bottlenecks to start with a smaller representation of the data, expands it for processing, and shrinks it again to reduce the number of computations. The model also removes non-linearities to maintain accuracy despite its simplified design. Like [MobileNet V1](./mobilenet_v1), it uses depthwise separable convolutions for efficiency.
```

> You can find checkpoints like [`google/mobilenet_v2_1.4_224`](https://huggingface.co/google/mobilenet_v2_1.4_224) (for classification) or [`google/deeplabv3_mobilenet_v2_1.0_513`](https://huggingface.co/google/deeplabv3_mobilenet_v2_1.0_513) (for segmentation) on the Hub. Check the [Google organization page](https://huggingface.co/google) for other variants. <!-- Consider linking to a dedicated Collection if one exists -->

Reviewer: ```suggestion
You can all the original MobileNet checkpoints under the [Google](https://huggingface.co/google?search_models=mobilenet) organization.
```

> [!TIP]
> Click on the MobileNet V2 models in the right sidebar for more examples of how to apply MobileNet to different vision tasks.

> The examples below demonstrate how to use MobileNetV2 for image classification and semantic segmentation with [`Pipeline`] or [`AutoModel`].

Reviewer: Lets keep it simple and just show one use case :)

```suggestion
The examples below demonstrate how to classify an image with [`Pipeline`] or the [`AutoModel`] class.
```

> **Image Classification**

Reviewer: ```suggestion
```

> <hfoptions id="usage-img-class">

Reviewer: Same comment for this `Pipeline` code snippet as `mobilenet_v1`

<hfoption id="Pipeline">

```python
import requests
from PIL import Image
from transformers import pipeline

# Use a pipeline as a high-level helper
pipe = pipeline("image-classification", model="google/mobilenet_v2_1.4_224") # Example checkpoint

# Load image from the web
url = "http://images.cocodataset.org/val2017/000000039769.jpg"
image = Image.open(requests.get(url, stream=True).raw)

# Pass image to the pipeline
results = pipe(image)
print(results)
# Example output: [{'label': 'tabby, tabby cat', 'score': 0.918...}, ...]
```

</hfoption>

> <hfoption id="AutoModel">

Reviewer: Apply same comment from `mobilenet_v1`

```python
import requests
import torch
from PIL import Image
from transformers import AutoImageProcessor, AutoModelForImageClassification

# Load processor and model
processor = AutoImageProcessor.from_pretrained("google/mobilenet_v2_1.4_224") # Example checkpoint
model = AutoModelForImageClassification.from_pretrained("google/mobilenet_v2_1.4_224")

# Load image from the web
url = "http://images.cocodataset.org/val2017/000000039769.jpg"
image = Image.open(requests.get(url, stream=True).raw)

# Preprocess image
inputs = processor(images=image, return_tensors="pt")

# Perform inference
with torch.no_grad():
    outputs = model(**inputs)
    logits = outputs.logits

# Post-process to get labels
predicted_label_id = logits.argmax(-1).item()
predicted_label = model.config.id2label[predicted_label_id]
print(predicted_label)
# Example output: tabby, tabby cat
```

</hfoption>
</hfoptions>

> **Semantic Segmentation**

Reviewer: We can remove the semantic segmentation example

from transformers import AutoImageProcessor, AutoModelForSemanticSegmentation

# Load processor and model
processor = AutoImageProcessor.from_pretrained("google/deeplabv3_mobilenet_v2_1.0_513") # Example checkpoint
model = AutoModelForSemanticSegmentation.from_pretrained("google/deeplabv3_mobilenet_v2_1.0_513")

# Load image from the web
url = "http://images.cocodataset.org/val2017/000000039769.jpg"
image = Image.open(requests.get(url, stream=True).raw)

# Preprocess image
inputs = processor(images=image, return_tensors="pt")

# Perform inference
with torch.no_grad():
    outputs = model(**inputs)
    logits = outputs.logits # shape (batch_size, num_labels, height, width)

# Post-process to get segmentation map
# Upsample logits to original image size
upsampled_logits = torch.nn.functional.interpolate(
    logits,
    size=image.size[::-1], # (height, width)
    mode='bilinear',
    align_corners=False
)
# Get predicted segmentation map
pred_seg = upsampled_logits.argmax(dim=1)[0]
print(pred_seg.shape)
# Example output: torch.Size([480, 640])
```

</hfoption>
</hfoptions>

<!-- Quantization - Not applicable -->
<!-- Attention Visualization - Not applicable for this model type -->

> ## Notes

Reviewer: Apply same comments from `mobilenet_v1`

-   **Checkpoint Naming:** Classification checkpoints often follow `mobilenet_v2_{depth_multiplier}_{resolution}`, like `mobilenet_v2_1.4_224`. Segmentation checkpoints (using DeepLabV3+ head) might have names like `deeplabv3_mobilenet_v2_{depth_multiplier}_{resolution}`.
-   **Variable Input Size:** Like V1, the model works with images of different sizes (minimum 32x32), handled by [`MobileNetV2ImageProcessor`].
-   **1001 Classes (Classification):** ImageNet-1k pretrained classification models output 1001 classes (index 0 is background).

> -   **Segmentation Head:** Segmentation models use a [DeepLabV3+](https://arxiv.org/abs/1802.02611) head, often pretrained on datasets like PASCAL VOC.

Reviewer: ```suggestion
-   The segmentation models use a [DeepLabV3+](https://huggingface.co/papers/1802.02611) head which is often pretrained on datasets like [PASCAL VOC](https://huggingface.co/datasets/merve/pascal-voc).
```

-   **Padding Differences:** Similar to V1, original TensorFlow checkpoints had dynamic padding. The HF PyTorch implementation uses static padding by default. Enable dynamic padding (TF behavior) via `tf_padding=True` in [`MobileNetV2Config`].
    ```python
    from transformers import MobileNetV2Config

    # Example: Load config with dynamic padding enabled
    config = MobileNetV2Config.from_pretrained("google/mobilenet_v2_1.4_224", tf_padding=True)
    ```
-   **Unsupported Features:**
    -   The HF implementation uses global average pooling, not the optional fixed 7x7 average pooling from the original paper.
    -   Extracting specific intermediate hidden states (e.g., from expansion layers 10/13) requires `output_hidden_states=True` (returning all states).
    -   For segmentation models, the final convolution layer of the backbone is computed even though the DeepLabV3+ head doesn't use it.

> ## Resources

Reviewer: Apply same comment from `mobilenet_v1`

import torch
from transformers import pipeline

pipeline = pipeline(
    task="image-classification",
    model="google/mobilenet_v2_1.4_224",
    torch_dtype=torch.float16,
    device=0
)
pipeline(images="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/pipeline-cat-chonk.jpeg")
```

</hfoption>
<hfoption id="AutoModel">

```python
import torch
import requests
from PIL import Image
from transformers import AutoModelForImageClassification, AutoImageProcessor

image_processor = AutoImageProcessor.from_pretrained(
    "google/mobilenet_v2_1.4_224",
)
model = AutoModelForImageClassification.from_pretrained(
    "google/mobilenet_v2_1.4_224",
)

url = "https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/pipeline-cat-chonk.jpeg"
image = Image.open(requests.get(url, stream=True).raw)
inputs = image_processor(image, return_tensors="pt")

with torch.no_grad():
  logits = model(**inputs).logits
predicted_class_id = logits.argmax(dim=-1).item()

class_labels = model.config.id2label
predicted_class_label = class_labels[predicted_class_id]
print(f"The predicted class label is: {predicted_class_label}")
```

</hfoption>
</hfoptions>

> <!-- Quantization - Not applicable -->
> <!-- Attention Visualization - Not applicable for this model type -->

Reviewer: ```suggestion
```

## Notes

> -   **Checkpoint Naming:** Classification checkpoints often follow `mobilenet_v2_{depth_multiplier}_{resolution}`, like `mobilenet_v2_1.4_224`. Segmentation checkpoints (using DeepLabV3+ head) might have names like `deeplabv3_mobilenet_v2_{depth_multiplier}_{resolution}`.

Reviewer: ```suggestion
-   Classification checkpoint names follow the pattern `mobilenet_v2_{depth_multiplier}_{resolution}`, like `mobilenet_v2_1.4_224`. `1.4` is the depth multiplier and `224` is the image resolution. Segmentation checkpoint names follow the pattern `deeplabv3_mobilenet_v2_{depth_multiplier}_{resolution}`.
```

> -   **Variable Input Size:** Like V1, the model works with images of different sizes (minimum 32x32), handled by [`MobileNetV2ImageProcessor`].

Reviewer: ```suggestion
-   While trained on images of a specific sizes, the model architecture works with images of different sizes (minimum 32x32). The [`MobileNetV2ImageProcessor`] handles the necessary preprocessing.
```

> -   **1001 Classes (Classification):** ImageNet-1k pretrained classification models output 1001 classes (index 0 is background).

Reviewer: ```suggestion
-   MobileNet is pretrained on [ImageNet-1k](https://huggingface.co/datasets/imagenet-1k), a dataset with 1000 classes. However, the model actually predicts 1001 classes. The additional class is an extra "background" class (index 0).
```

-   The segmentation models use a [DeepLabV3+](https://huggingface.co/papers/1802.02611) head which is often pretrained on datasets like [PASCAL VOC](https://huggingface.co/datasets/merve/pascal-voc).

> -   **Padding Differences:** Similar to V1, original TensorFlow checkpoints had dynamic padding. The HF PyTorch implementation uses static padding by default. Enable dynamic padding (TF behavior) via `tf_padding=True` in [`MobileNetV2Config`].

Reviewer: ```suggestion
-   The original TensorFlow checkpoints determines the padding amount at inference because it depends on the input image size. To use the native PyTorch padding behavior, set `tf_padding=False` in [`MobileNetV2Config`].
```

    ```python
    from transformers import MobileNetV2Config

>     # Example: Load config with dynamic padding enabled

Reviewer: ```suggestion
```

    config = MobileNetV2Config.from_pretrained("google/mobilenet_v2_1.4_224", tf_padding=True)
    ```

> -   **Unsupported Features:**
>     -   The HF implementation uses global average pooling, not the optional fixed 7x7 average pooling from the original paper.
>     -   Extracting specific intermediate hidden states (e.g., from expansion layers 10/13) requires `output_hidden_states=True` (returning all states).
>     -   For segmentation models, the final convolution layer of the backbone is computed even though the DeepLabV3+ head doesn't use it.

Reviewer: ```suggestion
-   The Transformers implementation does not support the following features.
    -   Uses global average pooling instead of the optional 7x7 average pooling with stride 2. For larger inputs, this gives a pooled output that is larger than a 1x1 pixel.
    -   `output_hidden_states=True` returns *all* intermediate hidden states. It is not possible to extract the output from specific layers for other downstream purposes.
    - Does not include the quantized models from the original checkpoints because they include "FakeQuantization" operations to unquantize the weights.
    -   For segmentation models, the final convolution layer of the backbone is computed even though the DeepLabV3+ head doesn't use it.
```
